pkg/kubelet/kubelet_server_journal.go

```diff
@@ -0,0 +1,372 @@
+	if len(fileLoggers) > 0 && previousBoot == 0 {
+		copyFileLogs(ctx, w, fileLoggers)
+	}
+
+	// To ensure we get boot logs in the Linux case
+	if len(nativeLoggers) == 0 && len(fileLoggers) == 0 {
+		a.copyServiceLogs(ctx, w, []string{}, previousBoot)
+	}
+}
+
+// splitNativeVsFileLoggers checks if each service logs to native OS logs or to a file and returns a list of services
+// that log natively vs maybe to a file
+func (a *journalArgs) splitNativeVsFileLoggers(ctx context.Context) ([]string, []string) {
+	var nativeLoggers []string
+	var fileLoggers []string
+
+	for _, service := range a.Services {
+		cmdStr, args := getServiceLogDestinationCmd(service)
+		cmd := exec.Command(cmdStr, args...)
+
+		go func() {
+			<-ctx.Done()
+			if p := cmd.Process; p != nil {
+				p.Kill()
+			}
+		}()
+
+		output, err := cmd.CombinedOutput()
+		// journalctl will not return an error even if the service is not using journald where as Get-WinEvent will.
+		if err != nil {
+			if _, ok := err.(*exec.ExitError); ok {
+				fileLoggers = append(fileLoggers, service)
+				continue
+			}
+		}
+
+		// Check the journalctl output to figure if the service is using journald or not. This is not needed in the
+		// Get-WinEvent case as the command returns an error if a service is not logging to the Application provider.
+		if checkForNativeLogger(output, service) {
+			nativeLoggers = append(nativeLoggers, service)
+		} else {
+			fileLoggers = append(fileLoggers, service)
+		}
+	}
+	return nativeLoggers, fileLoggers
+}
+
+// copyServiceLogs invokes journalctl or Get-WinEvent with the  provided args
+func (a *journalArgs) copyServiceLogs(ctx context.Context, w io.Writer, services []string, previousBoot int) {
+	cmdStr, args := getLoggingCmd(a, services, previousBoot)
+	cmd := exec.Command(cmdStr, args...)
+	cmd.Stdout = w
+	cmd.Stderr = w
+
+	// force termination
+	go func() {
+		<-ctx.Done()
+		if p := cmd.Process; p != nil {
+			p.Kill()
+		}
+	}()
+
+	if err := cmd.Run(); err != nil {
+		if _, ok := err.(*exec.ExitError); ok {
+			return
+		}
+		if previousBoot == 0 {
+			fmt.Fprintf(w, "error: journal output not available\n")
+		}
+	}
+}
+
+// copyFileLogs loops over all the services and attempts to collect the file logs of each service
+func copyFileLogs(ctx context.Context, w io.Writer, services []string) {
+	if ctx.Err() != nil {
+		return
+	}
+
+	for _, service := range services {
+		if len(service) > 0 {
+			heuristicsCat(ctx, w, service)
+		}
+	}
+}
+
+// heuristicsCat attempts to collect logs from either /var/log/service.log or /var/log/service/service.log or
+// /var/log/service*INFO or /var/log/service/service*INFO in that order stopping on first success.
+func heuristicsCat(ctx context.Context, w io.Writer, service string) {
+	logDir := "/var/log/"
+	logFiles := [4]string{logDir + service + ".log", logDir + service + "/" + service + ".log", logDir + service + "*INFO*",
+		logDir + service + "/" + service + "*INFO*"}
```

Comment:
On Windows (at least) having *INFO* instead of *INFO will likely result in getting duplicated logs.

KLog produces a file like `service.{machinename}.log.INFO.{timestamp}` each time a new logger is instantiated (like when kubelet restarts) and then creates a hardlink named `service.INFO` to the most recent file.

If the heuristics are `*INFO` this should always read the most recent log file and only the most recent log file.
If the heuristics are `*INFO*` we would get the contents of all previous log files but the contents of the most recent log file would be duplicated.

Each have advantages and disadvantages.
Either way we should make sure the comments match the code.

Reply:
Could we possibly look for `*INFO.*` or `*INFO?*` here first?
```suggestion
		logDir + service + "/" + service + "*INFO.*", logDir + service + "/" + service + "*INFO"}
```

Reply:
I've update it to first search for `*INFO.*` and then for `*INFO`. cc @aravindhp